src/librustc_mir/transform/promote_consts.rs

```diff
@@ -134,22 +156,604 @@ impl<'tcx> Visitor<'tcx> for TempCollector<'tcx> {
+                            | "cttz_nonzero"
+                            | "ctlz"
+                            | "ctlz_nonzero"
+                            | "wrapping_add"
+                            | "wrapping_sub"
+                            | "wrapping_mul"
+                            | "unchecked_shl"
+                            | "unchecked_shr"
+                            | "rotate_left"
+                            | "rotate_right"
+                            | "add_with_overflow"
+                            | "sub_with_overflow"
+                            | "mul_with_overflow"
+                            | "saturating_add"
+                            | "saturating_sub"
+                            | "transmute"
+                            => return Err(Unpromotable),
+
+                            _ => {}
+                        }
+                    }
+                    _ => {
+                        let is_const_fn =
+                            self.tcx.is_const_fn(def_id) ||
+                            self.tcx.is_unstable_const_fn(def_id).is_some() ||
+                            self.is_const_panic_fn(def_id);
+                        if !is_const_fn {
+                            return Err(Unpromotable);
+                        }
+                    }
+                }
+            }
+            _ => return Err(Unpromotable),
+        }
+
+        self.validate_operand(callee)?;
+        for arg in args {
+            self.validate_operand(arg)?;
+        }
+
+        Ok(())
+    }
+}
+
+pub fn validate_candidates(
+    tcx: TyCtxt<'tcx>,
+    body: &Body<'tcx>,
+    def_id: DefId,
+    temps: &IndexVec<Local, TempState>,
+    has_mut_interior: &BitSet<Local>,
+    needs_drop: &BitSet<Local>,
+    candidates: &[Candidate],
+) -> Vec<Candidate> {
+    let mut validator = Validator {
+        tcx,
+        param_env: tcx.param_env(def_id),
+        body,
+        is_static: false,
+        is_static_mut: false,
+        is_non_const_fn: false,
+        temps,
+        has_mut_interior,
+        needs_drop,
+
+        explicit: false,
+    };
+
+    // FIXME(eddyb) remove the distinctions that make this necessary.
+    let id = tcx.hir().as_local_hir_id(def_id).unwrap();
+    match tcx.hir().body_owner_kind(id) {
+        hir::BodyOwnerKind::Closure => validator.is_non_const_fn = true,
+        hir::BodyOwnerKind::Fn => {
+            if !tcx.is_const_fn(def_id) {
+                validator.is_non_const_fn = true;
+            }
+        },
+        hir::BodyOwnerKind::Static(hir::MutImmutable) => validator.is_static = true,
+        hir::BodyOwnerKind::Static(hir::MutMutable) => validator.is_static_mut = true,
+        _ => {}
+    }
+
+    candidates.iter().copied().filter(|&candidate| {
+        validator.explicit = match candidate {
+            Candidate::Ref(_) |
+            Candidate::Repeat(_) => false,
+            Candidate::Argument { .. } => true,
+        };
```

Comment:
Could we set these in `validate_candidate` instead of setting them here and asserting their values there?

Reply:
There was a subtle reason it was this weird but I don't remember now.
Maybe something to do with mutability?

Reply:
Yeah it's mutability.